docs/configuration.md
## Configuration
This bot can pull configuration from one of multiple locations. Each parameter will be parsed independently in the following order of precedence:
1. CLI arguments
    - See help menu by running `poglink -h` for more information.
2. Configuration File
    - Assumed to be named `config.yaml` within the data directory.
    - Data directory is set to `~/.poglink` when running natively unless otherwise specified. See [Docker Instructions](https://github.com/FM-17/poglink/blob/main/docs/docker-installation.md) for more information on how to configure this directory as a bind-mount.
    - See the supplied [sample configuration file](https://github.com/FM-17/poglink/blob/main/sample-config.yaml) for a reference of how to set parameters in this way.
3. Environment variables
    - Each parameter can be set via an environment variable prefixed with `BOT_`, in ALL CAPS.
    - E.g. to configure the bot's polling delay, set `BOT_POLLING_DELAY`.
4. Defaults (optional)
    - Some configuration parameters have default values assigned, which will be used
    in the absence of any other user-provided configuration values. 


### Parameter Summary
The following configuration parameters are available to be set in any of the above described methods:

| CLI Argument           | Env Var                  | Default                                   | Required | Description                                                          |
| ---------------------- | ------------------------ | ----------------------------------------- | -------- | -------------------------------------------------------------------- |
| `--allowed-roles`      | `BOT_ALLOWED_ROLES`      | None                                      | No       | Roles permitted to use bot commands (comma-separated list)           |
| `--polling-delay`      | `BOT_POLLING_DELAY`      | 60                                        | No       | Delay between each API check                                         |
| `--rates-urls`         | `BOT_RATES_URLS`         | [{"server_name": "Official","url": "https://cdn2.arkdedicated.com/asa/dynamicconfig.ini","color": "0x63BCC3",}] | Yes       | API URL to check for server rates (dict containing server title, url, and embed color)        |
| `--bans-url`           | `BOT_BANS_URL`           | http://arkdedicated.com/bansummary.txt    | No       | 🚧 [WIP] API URL to check for a ban summary                              |
| `--rates-channel-id`   | `BOT_RATES_CHANNEL_ID`   | None                                      | Yes      | Channel ID to post rates in                                          |
| `--bans-channel-id`    | `BOT_BANS_CHANNEL_ID`    | None                                      | Yes      | 🚧 [WIP] Channel ID to post ban summary in  WIP                                   |
| `--token `             | `BOT_TOKEN`              | None                                      | Yes      | Bot token (from Discord Developer Portal)                            |
| `--data-dir`           | `BOT_DATA_DIR`           | ~/.poglink                        | No       | Directory that will contain bot data, such as the `config.yaml` file |
| `--debug`              | `BOT_DEBUG`              | False                                     | No       | Enables debug logging |
| `--send-embed-on-startup` | `BOT_SEND_EMBED_ON_STARTUP`| False                                | No       | Send embed containing current rates found after starting the bot for the first time. | 

## Configuration Examples
- - - 
**Example 1:** Running Poglink to monitor rates on Official PC and Smalltribes servers

### Using `config.yaml`
```yaml
# REQUIRED 
token: JIMyUDY3Lah3SDd5JAM3Xds1.UH15df.lgOyDi5al5D_7d21gDDkAdHdlT3 # fake token
rates_channel_id: 733415533152301503 

rates_urls: 
- server_name: Official
  url:  https://cdn2.arkdedicated.com/asa/dynamicconfig.ini
  color: 0x63BCC3
- server_name: Smalltribes
  url: https://cdn2.arkdedicated.com/asa/smalltribes_dynamicconfig.ini
  color: 0xF8DE74
```

- - - 
**Example 2:** Running Poglink from a custom data directory

### Using `config.yaml` 

> ⚠ You cannot set a custom data directory solely from the `config.yaml` file because the bot requires the data directory location in order to find the `config.yaml` file.

### Using the CLI
```bash
poglink --data-dir=~/custom_data_dir
```


poglink/config.py
import logging
import os

import yaml

from poglink.error import ConfigReadError
from poglink.utils import parse_bool, parse_list

logger = logging.getLogger(__name__)

MIN_POLLING_DELAY = 5

DEFAULT_CONFIG = {
    "allowed_roles": [],
    "polling_delay": 60,
    "rates_urls": [
        {
            "server_name": "Official",
            "url": "https://cdn2.arkdedicated.com/asa/dynamicconfig.ini",
            "color": "0x63BCC3",
        }
    ],
    # "bans_url": "http://arkdedicated.com/bansummary.txt",
    "rates_channel_id": None,
    # "bans_channel_id": None, # TODO: Reimplement when bans are enabled
    "token": None,
    "data_dir": "~/.poglink",
    "send_embed_on_startup": False,
}

REQUIRED_VALUES = [
    "token",
    "rates_channel_id",
    # "bans_channel_id", # TODO: Reimplement when bans are enabled
]

LIST_VALUES = [
    "allowed_roles",
]

BOOLEAN_VALUES = ["send_embed_on_startup"]


def setup_config(args, default_config=DEFAULT_CONFIG):
    # Attempt to load config values from file if provided
    data_dir = os.path.expanduser(
        args.data_dir or os.getenv("BOT_DATA_DIR") or DEFAULT_CONFIG.get("data_dir")
    )
    config_path = os.path.join(data_dir, "config.yaml")
    logger.debug(f"Setting up configuration. Checking for config file: {config_path}")

    if os.path.exists(config_path):
        try:
            with open(os.path.expanduser(config_path)) as f:
                config_from_file = yaml.safe_load(f)
        except Exception as e:
            logger.error(f"Problem reading configuration file at {config_path}: {e}")
            raise ConfigReadError(e) from e
    else:
        logger.warning(
            f"No configuration file found at {config_path}. Configuration must be set via CLI args or environment variables."
        )
        config_from_file = {}

    # For each configuration value, attempt to obtain value in the specified order of priority:
    config = {}
    for key, default_val in default_config.items():
        config[key] = (
            getattr(args, key)
            or config_from_file.get(key)
            or os.getenv(f"BOT_{key.upper()}")
            or default_val
        )

    # handle special cases for boolean values
    for param in BOOLEAN_VALUES:
        if not isinstance(config[param], bool):
            config[param] = parse_bool(config[param])

    # handle special cases for list parsing
    for val in LIST_VALUES:
        if isinstance(config[val], str):
            config[val] = parse_list(config[val])
        elif isinstance(config[val], list):
            pass
        else:
            logger.warning(
                f"Incorrect variable format for {val}; should be comma separated list. Actual value: {config[val]}"
            )

    # handle special case for polling delay
    if config["polling_delay"] < MIN_POLLING_DELAY:
        logger.warning(
            f"Polling delay ({config['polling_delay']}) below minimum value. Setting to minimum value ({MIN_POLLING_DELAY})."
        )
        config["polling_delay"] = MIN_POLLING_DELAY

    logger.debug(f"Configuration loaded from sources: {config}")
    return config
